modules/apps/document-library/document-library-test/src/testIntegration/java/com/liferay/document/library/app/service/test/DLAppServiceWhenCopyingAFileShortcutTest.java

```diff
@@ -0,0 +1,189 @@
+				group.getGroupId(), TestPropsValues.getUserId()));
+
+		targetParentFolder = DLAppServiceUtil.addFolder(
+			null, targetGroup.getGroupId(),
+			DLFolderConstants.DEFAULT_PARENT_FOLDER_ID, "Target Test Folder",
+			RandomTestUtil.randomString(),
+			ServiceContextTestUtil.getServiceContext(
+				targetGroup.getGroupId(), TestPropsValues.getUserId()));
+	}
+
+	@Test
+	public void testShouldSucceedFromHomeToHomeInDifferentSite()
+		throws Exception {
+
+		_copyFromHomeToHome(group.getGroupId(), targetGroup.getGroupId());
+	}
+
+	@Test
+	public void testShouldSucceedFromHomeToSubfolder() throws Exception {
+		_copyFromHome(group.getGroupId(), newParentFolder);
+	}
+
+	@Test
+	public void testShouldSucceedFromHomeToSubfolderInDifferentSite()
+		throws Exception {
+
+		_copyFromHome(group.getGroupId(), targetParentFolder);
+	}
+
+	@Test
+	public void testShouldSucceedFromSubfolderToHome() throws Exception {
+		_copyToHome(parentFolder, group.getGroupId());
+	}
+
+	@Test
+	public void testShouldSucceedFromSubfolderToHomeInDifferentSite()
+		throws Exception {
+
+		_copyToHome(parentFolder, targetGroup.getGroupId());
+	}
+
+	@Test
+	public void testShouldSucceedFromSubfolderToSubfolder() throws Exception {
+		_copy(parentFolder, newParentFolder);
+	}
+
+	@Test
+	public void testShouldSucceedFromSubfolderToSubfolderInDifferentSite()
+		throws Exception {
+
+		_copy(parentFolder, targetParentFolder);
+	}
+
+	@Test
+	public void testShouldSucceedIfDestinationIsSameHome() throws Exception {
+		_copyFromHomeToHome(group.getGroupId(), group.getGroupId());
+	}
+
+	@Test
+	public void testShouldSucceedIfDestinationIsSameSubfolder()
+		throws Exception {
+
+		_copy(parentFolder, parentFolder);
+	}
+
+	protected FileEntry fileEntry;
+	protected Folder newParentFolder;
+	protected Folder targetParentFolder;
+
+	private void _copy(Folder sourceParentFolder, Folder targetParentFolder)
+		throws Exception {
+
+		FileShortcut fileShortcut = DLAppServiceUtil.addFileShortcut(
+			sourceParentFolder.getRepositoryId(),
+			sourceParentFolder.getFolderId(), fileEntry.getFileEntryId(),
+			ServiceContextTestUtil.getServiceContext(
+				sourceParentFolder.getGroupId()));
+
+		DLAppServiceUtil.copyFileShortcut(
+			fileShortcut.getFileShortcutId(), targetParentFolder.getFolderId(),
+			targetParentFolder.getGroupId(),
+			ServiceContextTestUtil.getServiceContext(
+				targetParentFolder.getGroupId()));
+	}
+
+	private void _copyFromHome(long sourceGroupId, Folder targetParentFolder)
+		throws Exception {
+
+		FileShortcut fileShortcut = DLAppServiceUtil.addFileShortcut(
+			sourceGroupId, DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+			fileEntry.getFileEntryId(),
+			ServiceContextTestUtil.getServiceContext(sourceGroupId));
+
+		DLAppServiceUtil.copyFileShortcut(
+			fileShortcut.getFileShortcutId(), targetParentFolder.getFolderId(),
+			targetParentFolder.getGroupId(),
+			ServiceContextTestUtil.getServiceContext(
+				targetParentFolder.getGroupId()));
+	}
```

Comment:
Same as my previous comment in the other test file. Remove duplication by adding a `copy(long, long, long, long)` method (and override for convenience if necessary).